trestle/core/commands/remove.py

```diff
@@ -15,10 +15,113 @@
+        parent_model, parent_alias = fs.get_contextual_model_type(file_path.absolute())
+        parent_object = parent_model.oscal_read(file_path.absolute())
+        parent_element = Element(parent_object, utils.classname_to_alias(parent_model.__name__, 'json'))
+
+        add_plan = Plan()
+
+        # Do _remove for each element_path specified in args
+        element_paths: list[str] = args[const.ARG_ELEMENT].split(',')
+        for elm_path_str in element_paths:
+            element_path = ElementPath(elm_path_str)
+            remove_action, parent_element = self.remove(element_path, parent_model, parent_element)
+            add_plan.add_action(remove_action)
+
+        create_action = CreatePathAction(file_path.absolute(), True)
+        write_action = WriteFileAction(
+            file_path.absolute(), parent_element, FileContentType.to_content_type(file_path.suffix)
+        )
+        add_plan.add_action(remove_action)
+        add_plan.add_action(create_action)
+        add_plan.add_action(write_action)
+        add_plan.simulate()
+        add_plan.execute()
+
+    @classmethod
+    def remove(cls, element_path, parent_model, parent_element):
+        """For the element_path, remove a model from the parent_element of a given parent_model.
+
+        First we check if there is an existing element at that path
+        If not, we complain.
+        Then we set up an action plan to update the model (specified by file_path) in memory,
+        return the action and return the parent_element.
+
+        LIMITATIONS:
+        1. This does not remove elements of a list or dict. Instead, the entire list or dict is removed.
+        2. This cannot remove arbitrarily named elements that are not specified in the schema.
+        For example, "responsible-parties" contains named elements, e.g., "organisation". The tool will not
+        remove the "organisation" as it is not in the schema, but one can remove its elements, e.g., "party-uuids".
+        """
+        element_path_list = element_path.get_full_path_parts()
+        if '*' in element_path_list:
+            raise err.TrestleError('trestle remove does not support Wildcard element path.')
+
+        try:
+            deleting_element = parent_element.get_at(element_path)
+
+            if deleting_element is not None:
+                # The element already exists
+                if type(deleting_element) is list:
+                    pass
+                    warnings.warn('trestle remove does not support removing elements of a list', Warning)
+                    warnings.warn('trestle remove of a list removes the entire list', Warning)
+                elif type(deleting_element) is dict:
+                    pass
+                    warnings.warn('trestle remove does not support removing dict elements', Warning)
+                    warnings.warn('trestle remove of a dict element removes the entire dict element', Warning)
```

Comment:
Warnings give a strange user experience. Use logger.warn(). Not that trestle has a logging handler which is centralised using trestle.utils.log.get_logger()

Reply:
```
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.backmatter
Remove failed: Bad element path: catalog.backmatter
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata
Remove failed: Last stream position is not available to rollback to
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata.*
Remove failed: trestle remove does not support Wildcard element path.
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$
```

The handling of a validation failure from a plan simulation is not ideal, but the actual error, i.e., removing a required element in this case, is not properly passed back by pydantic, which normally throws:
```
(trestle) [jeffetan@Jeffs-MacBook-Pro compliance-trestle (feature/66-remove)]$ python /Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/tr
estle/__main__.py remove -f ./jeff-test/catalogs/minimal_catalog.json --element catalog.metadata
Traceback (most recent call last):
  File "/Users/jeffetan/Git/src/github.com/IBM/compliance-trestle/trestle/core/models/elements.py", line 359, in set_at
    setattr(preceding_elm, sub_element_name, model_obj)
  File "pydantic/main.py", line 394, in pydantic.main.BaseModel.__setattr__
pydantic.error_wrappers.ValidationError: 1 validation error for Catalog
metadata
  none is not an allowed value (type=type_error.none.not_allowed)
```

So for now it's not a useful error message from simulate(). I do note that it comes from simulate() in the log, but not in the final error returned.

And yes, I am now using a logger instead of warnings.